OPM/opm-rule-onboarding.ps1

```diff
@@ -0,0 +1,540 @@
+	} catch {
+		Log-Msg -Type Error -MSG "Could not change SSL settings to use TLS 1.2"
+		Log-Msg -Type Error -MSG (Collect-ExceptionMessage $_.Exception) -ErrorAction "SilentlyContinue"
+	}
+}
+
+
+# Check that the PVWA URL is OK
+If (![string]::IsNullOrEmpty($PVWAURL))
+{
+	If ($PVWAURL.Substring($PVWAURL.Length-1) -eq "/")
+	{
+		$PVWAURL = $PVWAURL.Substring(0,$PVWAURL.Length-1)
+	}
+	
+	try{
+		# Validate PVWA URL is OK
+		Log-Msg -Type Debug -MSG  "Trying to validate URL: $PVWAURL"
+		Invoke-WebRequest -UseBasicParsing -DisableKeepAlive -Uri $PVWAURL -Method 'Head' -TimeoutSec 30 | Out-Null
+	} catch [System.Net.WebException] {
+		If(![string]::IsNullOrEmpty($_.Exception.Response.StatusCode.Value__))
+		{
+			Log-Msg -Type Error -MSG $_.Exception.Response.StatusCode.Value__
+		}
+	}
+	catch {		
+		Log-Msg -Type Error -MSG "PVWA URL could not be validated"
+		Log-Msg -Type Error -MSG (Collect-ExceptionMessage $_.Exception) -ErrorAction "SilentlyContinue"
+	}
+	
+}
+else
+{
+	Log-Msg -Type Error -MSG "PVWA URL cannot be empty."
+	exit
+}
+
+Log-Msg -Type Info -MSG "Getting PVWA Credentials to start Onboarding OPM Rules" -SubHeader
+
+
+#region [Logon]
+	# Get Credentials to Login
+	# ------------------------
+	$caption = "OPM Rule Onboarding Utility"
+	$msg = "Enter your User name and Password"; 
+	$creds = $Host.UI.PromptForCredential($caption,$msg,"","")
+	if ($creds -ne $null)
+	{
+		if($AuthType -eq "radius" -and ![string]::IsNullOrEmpty($OTP))
+		{
+			$g_LogonHeader = $(Get-LogonHeader -Credentials $creds -RadiusOTP $OTP)
+		}
+		else
+		{
+			$g_LogonHeader = $(Get-LogonHeader -Credentials $creds)
+		}
+	}
+	else { 
+		Log-Msg -Type Error -MSG "No credentials were entered." -Footer
+		exit
+	}
+#endregion
+
+#region [Read Accounts CSV file and Create Accounts]
+	If([string]::IsNullOrEmpty($CsvPath))
+	{
+		$CsvPath = OpenFile-Dialog($g_CsvDefaultPath)
+	}
+	$delimiter = $(If ($CsvDelimiter -eq "Comma") { "," } else { "`t" } )
+	$rulesCSV = Import-CSV $csvPath -Delimiter $delimiter
+	$rowCount = $($rulesCSV.Safe.Count)
+	$counter = 0
+	$successCounter = 0
+	Log-Msg -Type Info -MSG "Starting to Onboard $rowCount rules." -SubHeader
+	ForEach ($rule in $rulesCSV)
+	{
+		if ($null -ne $rule)
+		{
+			try {
+				# Check mandatory fields
+				If([string]::IsNullOrEmpty($rule.PlatformId)) { throw "Missing mandatory field: Platform ID" }
+				If([string]::IsNullOrEmpty($rule.Command)) { throw "Missing mandatory field: Command" }
+				If([string]::IsNullOrEmpty($rule.UserName)) { throw "Missing mandatory field: UserName" }
+				
+				$rule.PlatformId = (Get-TrimmedString $rule.PlatformID)
+				$objRule = "" | Select-Object "Command", "CommandGroup", "PermissionType", "Restrictions", "UserName"
+				$objRule.Command = (Get-TrimmedString $rule.Command)
+				$objRule.CommandGroup = (Get-TrimmedString $rule.CommandGroup)
+				$objRule.PermissionType = (Get-TrimmedString $rule.PermissionType)
```

Comment:
Here for example try to think if NULL is a valid type
Maybe you want to make the Get-TrimmedString function a bit more clever by adding a 'default if null' output to avoid having null values and setting defaults in advance

Reply:
Yes, this is a good idea. For the first iteration of the script, I provided very little input validation. Essentially, we have the minimum of validations on required fields and what the inputs can be based on documentation. I listed some limitations at the top of the script which I would intend to improve upon in a future iteration of the script.